### ISSUE TITLE
(sig)STOPing and (sig)CONTinuing a send|receive pipe causes: ERROR: crc32 mismatch in command

### ISSUE THREAD

**Issue by calestyo** [kdave/btrfs-progs#719]
Hey.

From [[0](https://lore.kernel.org/linux-btrfs/6f340d377c584239a1b8d4236d404d708e61d968.camel@scientia.org/)]:

I have experienced this myself already some years ago and apparently
others[[1](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=888858
)], too.

When stopping a send|receive pipeline (both on the same host, i.e. no
ssh or so involved), and continuing it again, the whole thing dies
immediately with an:
   ERROR: crc32 mismatch in command

Which is particularly annoying when one wanted to make a backup of 9TB
and was nearly finished and had forgotten that this breaks. ;-)

Any way to have that fixed?

Cheers,
Chris.

PS: kernel 6.5.10

**Comment by calestyo** [comment]
btw: It's clear that one cannot catch `SIGSTOP`, but if the problem itself (i.e. that the fails after `SIGCONT` would be unfixable, then maybe one could do something to at least prevent from accidental `SIGSTOP`, which may happen as easy as doing a `Ctrl-Z` in the terminal... I thought about temporarily disabling the `stop CHAR`, like via `stty susp ''`.

**Comment by CyberShadow** [comment]
I suspect that this is a regression; has anyone tried older kernels or btrfs-progs versions?

**Comment by Zygo** [comment]
...or somewhere there's a missing check for a restarted syscall

**Comment by kdave** [comment]
Yes, it must be the unsafe (wrt EINTR) read. It's not handled anywhere.

**Comment by kdave** [comment]
I can reproduce it but I don't understand what's wrong. It's not missing EINTR, because that would return an error with a different message (receive.c:read_buf). Neither handling SIGCONT with SA_RESTART works.

**Comment by kdave** [comment]
Also tested receive with 4.8.5 f6f9cd0cb76f52febd4f6579a4fb176c4d501068 which was last that did manual buffer copy instead of splice (also with SA_RESTART).

**Comment by kdave** [comment]
I think it's broken in kernel. Running just the send side (with SA_RESTART) as `btrfs send | cat > stream`, Ctrl-Z, fg, Ctrl-C and then running `btrfs receive --dump < stream` prints the known errors like the "crc32c mismatch" or "ERROR: command length 1718776930 too big for buffer 65536", while the expected error would be a short read or incomplete command.
